### PR TITLE
Raise an error when resuming training with Apex

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Alternative approach to  #12389

Fixes https://github.com/Lightning-AI/lightning/issues/11488

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @justusschock @carmocca @awaelchli @akihironitta @rohitgr7

### CODE REVIEW

File: src/pytorch_lightning/plugins/precision/apex_amp.py [hunk 1]
```diff
@@ -97,6 +97,3 @@ def optimizer_step(
 
     def state_dict(self) -> Dict[str, Any]:
         return amp.state_dict()
-
-    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
-        amp.load_state_dict(state_dict)
```
Comment (awaelchli) [inline src/pytorch_lightning/plugins/precision/apex_amp.py:102]: By deleting this, we do nothing now (super implementation) for all stages by default.

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -209,6 +209,10 @@ def restore_training_state(self) -> None:
     def restore_precision_plugin_state(self) -> None:
         """Restore the precision plugin state from the pre-loaded checkpoint."""
         prec_plugin = self.trainer.precision_plugin
+
+        if isinstance(prec_plugin, ApexMixedPrecisionPlugin) and self.trainer.state.fn == TrainerFn.FITTING:
+            raise RuntimeError("Resuming training with APEX is currently not supported.")
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:214]: This only matters in training. We can still resume via trainer.test(), but optimizer states don't matter there.

File: tests/tests_pytorch/models/test_amp.py [hunk 3]
```diff
@@ -211,3 +211,24 @@ def configure_optimizers(self):
...
     assert isinstance(trainer.lr_scheduler_configs[0].scheduler.optimizer, optim.Adam)
     assert isinstance(trainer.lr_scheduler_configs[1].scheduler.optimizer, optim.SGD)
+
+
+@RunIf(min_cuda_gpus=1, amp_apex=True)
+def test_amp_with_apex_reload(tmpdir):
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_steps=5,
```
Comment (carmocca) [inline tests/tests_pytorch/models/test_amp.py:221]: Why that many steps?

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 4]
```diff
@@ -209,6 +209,10 @@ def restore_training_state(self) -> None:
     def restore_precision_plugin_state(self) -> None:
         """Restore the precision plugin state from the pre-loaded checkpoint."""
         prec_plugin = self.trainer.precision_plugin
+
+        if isinstance(prec_plugin, ApexMixedPrecisionPlugin) and self.trainer.state.fn == TrainerFn.FITTING:
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:213]: What about a `Trainer` without apex enabled loaded with a checkpoint that has apex state? That doesn't raise. Should it?

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 4 above)
Comment (justusschock) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:213]: Why not raise this error within the `ApexMixedPrecisionPlugin`? That way it would be encapsulated within the responsible plugin.

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 4 above)
Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:213]: Because inside there, we don't know whether the trainer is training or testing. 
I don't know if I can move it inside like you said. There would still have to be an isinstance check somewhere to set this special attribute.

Comment (justusschock) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:213]: I imagined it to be something like this:


```python
def load_state_dict(self, state_dict):
    self._state_dict_loaded = True
    return super().load_state_dict(state_dict)

def optimizer_step(self, *args, **kwargs): # args and kwargs for simplicity, ofc keep real signature
    # in here we definitely know, that we are training
   if self._state_dict_loaded:
        raise MisconfigurationException(...)
   ... # continue with usual optimizer step.

```

Wouldn't that work?

Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py]: I understand what you mean now. Yes, I definitely like this more!

File: src/pytorch_lightning/plugins/precision/apex_amp.py [hunk 5]
```diff
@@ -83,6 +84,8 @@ def optimizer_step(
         closure: Callable[[], Any],
         **kwargs: Any,
     ) -> Any:
+        if self._state_dict_loaded:
+            raise RuntimeError("Resuming training with APEX is currently not supported.")
```
Comment (rohitgr7) [inline src/pytorch_lightning/plugins/precision/apex_amp.py:88]: should we suggest not using apex while configuring the Trainer to avoid this?

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/plugins/precision/apex_amp.py (hunk 5 above)
Comment (awaelchli) [reply src/pytorch_lightning/plugins/precision/apex_amp.py:88]: How would you show this error message during Trainer configuration? What would be the condition?

Comment (rohitgr7) [reply src/pytorch_lightning/plugins/precision/apex_amp.py:88]:
```py
raise RuntimeError("Resuming training with APEX is currently not supported. Please configure your `Trainer(amp_backend=None)` while resuming.")
```

Comment (rohitgr7) [reply src/pytorch_lightning/plugins/precision/apex_amp.py:88]: oh! my comment might have confused you. I meant let users know how they can update the Trainer configuration to resume without apex :)
